cli/main.rs

```diff
@@ -559,42 +619,49 @@ async fn run_from_stdin(flags: Flags) -> Result<(), AnyError> {
 }
 
 async fn run_with_watch(flags: Flags, script: String) -> Result<(), AnyError> {
-  let main_module = ModuleSpecifier::resolve_url_or_path(&script)?;
-  let program_state = ProgramState::new(flags.clone())?;
+  let module_resolver = || {
+    let script = script.clone();
+    let flags = flags.clone();
+    async move {
+      let main_module = ModuleSpecifier::resolve_url_or_path(&script)?;
+      let program_state = ProgramState::new(flags)?;
+      let handler = Rc::new(RefCell::new(FetchHandler::new(
+        &program_state,
+        Permissions::allow_all(),
+      )?));
+      let mut builder = module_graph::GraphBuilder::new(
+        handler,
+        program_state.maybe_import_map.clone(),
+        program_state.lockfile.clone(),
+      );
+      builder.add(&main_module, false).await?;
+      let module_graph = builder.get_graph();
+
+      // Find all local files in graph
+      let mut paths_to_watch: Vec<PathBuf> = module_graph
+        .get_modules()
+        .iter()
+        .filter_map(|specifier| specifier.as_url().to_file_path().ok())
+        .collect();
+
+      if let Some(import_map) = program_state.flags.import_map_path.as_ref() {
+        paths_to_watch
+          .push(fs::resolve_from_cwd(std::path::Path::new(import_map))?);
+      }
 
-  let handler = Rc::new(RefCell::new(FetchHandler::new(
-    &program_state,
-    Permissions::allow_all(),
-  )?));
-  let mut builder = module_graph::GraphBuilder::new(
-    handler,
-    program_state.maybe_import_map.clone(),
-    program_state.lockfile.clone(),
-  );
-  builder.add(&main_module, false).await?;
-  let module_graph = builder.get_graph();
-
-  // Find all local files in graph
-  let mut paths_to_watch: Vec<PathBuf> = module_graph
-    .get_modules()
-    .iter()
-    .filter(|specifier| specifier.as_url().scheme() == "file")
-    .map(|specifier| specifier.as_url().to_file_path().unwrap())
-    .collect();
-
-  if let Some(import_map) = program_state.flags.import_map_path.clone() {
-    paths_to_watch
-      .push(fs::resolve_from_cwd(std::path::Path::new(&import_map)).unwrap());
-  }
+      Ok((paths_to_watch, main_module))
+    }
+    .boxed_local()
+  };
 
-  // FIXME(bartlomieju): new file watcher is created on after each restart
-  file_watcher::watch_func(&paths_to_watch, move || {
-    // FIXME(bartlomieju): ProgramState must be created on each restart - otherwise file fetcher
-    // will use cached source files
-    let gs = ProgramState::new(flags.clone()).unwrap();
-    let permissions = Permissions::from_flags(&flags);
-    let main_module = main_module.clone();
+  let operation = |main_module: ModuleSpecifier| {
+    let flags = flags.clone();
     async move {
+      let permissions = Permissions::from_flags(&flags);
+      // FIXME(bartlomieju): ProgramState must be created on each restart - otherwise file fetcher
+      // will use cached source files
```

Comment:
This problem should be fixed by https://github.com/denoland/deno/pull/8328, we must be careful on the timing between the two PRs

Reply:
I tried to check if this is still the case or not, but it seemed impossible to create `ProgramState` only once before the watcher starts, because of lifetime and trait bound restriction.
So effectively `ProgramState` ended up being created every time the watcher restarts. I think it's okay, so just removed the FIXME comment.